library/src/scala/runtime/stdLibPatches/Predef.scala

```diff
@@ -46,7 +46,8 @@ object Predef:
    *  }}}
    */
   extension [T](x: T | Null) inline def nn: x.type & T =
-    scala.runtime.Scala3RunTime.nn(x)
+    if x.asInstanceOf[Any] == null then scala.runtime.Scala3RunTime.nnFail()
+    x.asInstanceOf[x.type & T]
```

Comment:
Did you try
```suggestion
    x.getClass() // throws NPE if x is null, no-op otherwise
    x.asInstanceOf[x.type & T]
```
I've seen javac emit dummy calls to ` getClass()` like this for the purpose of throwing NPEs. It might be even more efficient if the JVM recognizes the pattern.

Reply:
Yes, that's probably a good idea.

Reply:
Can we find out

 - what error message gets emitted for a `x.getClass` call if `x` is null?
 - whether javac in later versions still emits the `getClass` or whether it does something else?

It would be good to do the same thing as javac since that's probably the smallest and fastest.

Reply:
I wonder if the JVM does some better optimization for `java.util.Objects.requireNonNull`.

Reply:
I believe `requireNonNull` is not the best we can do since it requires a useless cast after the call. We want to do the fastest possible non-null check (which should return Unit), followed by the cast to the target type.